src/OpenTelemetry/Metrics/Exemplar/Exemplar.cs

```diff
@@ -28,31 +30,149 @@ namespace OpenTelemetry.Metrics;
 #endif
     struct Exemplar
 {
+    internal HashSet<string>? KeyFilter;
+    private int tagCount;
+    private KeyValuePair<string, object?>[]? tagStorage;
+    private MetricPointValueStorage valueStorage;
+    private int isCriticalSectionOccupied;
+
     /// <summary>
     /// Gets the timestamp.
     /// </summary>
-    public DateTimeOffset Timestamp { get; internal set; }
+    public DateTimeOffset Timestamp { get; private set; }
 
     /// <summary>
     /// Gets the TraceId.
     /// </summary>
-    public ActivityTraceId? TraceId { get; internal set; }
+    public ActivityTraceId? TraceId { get; private set; }
 
     /// <summary>
     /// Gets the SpanId.
     /// </summary>
-    public ActivitySpanId? SpanId { get; internal set; }
+    public ActivitySpanId? SpanId { get; private set; }
 
-    // TODO: Leverage MetricPointValueStorage
-    // and allow double/long instead of double only.
+    /// <summary>
+    /// Gets the long value.
+    /// </summary>
+    public long LongValue
+    {
+        readonly get => this.valueStorage.AsLong;
+        private set => this.valueStorage.AsLong = value;
+    }
 
     /// <summary>
     /// Gets the double value.
     /// </summary>
-    public double DoubleValue { get; internal set; }
+    public double DoubleValue
+    {
+        readonly get => this.valueStorage.AsDouble;
+        private set => this.valueStorage.AsDouble = value;
+    }
 
     /// <summary>
     /// Gets the FilteredTags (i.e any tags that were dropped during aggregation).
     /// </summary>
-    public List<KeyValuePair<string, object?>>? FilteredTags { get; internal set; }
+    public readonly ReadOnlyFilteredTagCollection FilteredTags
+        => new(this.KeyFilter, this.tagStorage ?? Array.Empty<KeyValuePair<string, object?>>(), this.tagCount);
+
+    internal void Update<T>(in ExemplarMeasurement<T> measurement)
+        where T : struct
+    {
+        if (Interlocked.Exchange(ref this.isCriticalSectionOccupied, 1) != 0)
+        {
+            // Some other thread is already writing, abort.
+            return;
+        }
+
+        this.Timestamp = DateTimeOffset.UtcNow;
+
+        if (typeof(T) == typeof(long))
+        {
+            this.LongValue = (long)(object)measurement.Value;
+        }
+        else if (typeof(T) == typeof(double))
+        {
+            this.DoubleValue = (double)(object)measurement.Value;
+        }
+        else
+        {
+            Debug.Fail("Invalid value type");
+            this.DoubleValue = Convert.ToDouble(measurement.Value);
```

Comment:
Any need to do this `ToDouble` here?

Reply:
It is really just a quirk of the generics. This code won't ever execute unless someone messes up during dev and tries to call update with something other than `double` or `long` (eg `exemplar.Update(new ExemplarMeasurement<decimal>(...))`). I put the `Debug.Fail` in there to hopefully alert that future dev that they made a mistake 😄 When the JIT hits this method it should special-case things and elide/remove the invalid paths.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/Metric.cs#L155 This should prevent any unsupported types at the instrument creation itself right?